File: go/989.go
func addToArrayForm(A []int, K int) []int {
    var B []int
    for ; K != 0; K /= 10 {
        B = append(B, K%10)
    }
    for size, i := len(B), 0; i < size/2; i++ {
        B[i], B[size-1-i] = B[size-i-1], B[i]
    }
    carry, s1, s2, ret := 0, len(A)-1, len(B)-1, make([]int, 0)
    for ; s1 >= 0 && s2 >= 0; s1, s2 = s1-1, s2-1 {
        temp := carry + A[s1] + B[s2]
        ret = append(ret, temp%10)
        carry = temp / 10
    }
    for ; s1 >= 0; s1-- {
        temp := carry + A[s1]
        ret = append(ret, temp%10)
        carry = temp / 10
    }
    for ; s2 >= 0; s2-- {
        temp := carry + B[s2]
        ret = append(ret, temp%10)
        carry = temp / 10
    }
    if carry > 0 {
        ret = append(ret, carry)
    }
    for size, i := len(ret), 0; i < size/2; i++ {
        ret[i], ret[size-i-1] = ret[size-i-1], ret[i]
    }
    return ret
}

File: cpp/989.cpp
class Solution {
public:
    vector<int> addToArrayForm(vector<int>& A, int K) {
        vector<int> B;
        for (; K != 0; K /= 10) {
            B.push_back(K%10);
        }
        for (int i = 0, size = B.size(); i < size / 2; i++) 
            swap(B[i], B[size-i-1]);
        int i = A.size() - 1, j = B.size() - 1, carry = 0;
        vector<int> ret;
        while (i >= 0 && j >= 0) {
            int temp = carry + A[i--] + B[j--];
            ret.push_back(temp%10);
            carry = temp / 10;
        }
        while (i >= 0) {
            int temp = carry + A[i--];
            ret.push_back(temp%10);
            carry = temp / 10;
        }
        while (j >= 0) {
            int temp = carry + B[j--];
            ret.push_back(temp%10);
            carry = temp / 10;
        }
        if (carry)
            ret.push_back(carry);
        for (int i = 0, size = ret.size(); i < size / 2; i++) 
            swap(ret[i], ret[size-i-1]);
        return ret;
    }
};
